dra-evolution/pkg/api/capacity_types.go

```diff
@@ -3,168 +3,110 @@ package api
-	Name string
-	// Conditions contains the latest observation of the device's state.
-	Conditions []metav1.Condition `json:"conditions"`
+	// DriverName identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only ResourcePool
+	// objects with a certain driver name.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	DriverName string `json:"driverName" protobuf:"bytes,3,name=driverName"`
+
+	// Devices lists all available devices in this pool.
+	Devices []Device `json:"devices,omitempty"`
 
-	// ClaimUIDs contains the UIDs of the claims to which this device
-	// is allocated.
-	ClaimUIDs []types.UID
+	// FUTURE EXTENSION: some other kind of list, should we ever need it.
+	// Old clients seeing an empty Devices field can safely ignore the (to
+	// them) empty pool.
 }
 
-// Device is used to track individual devices in a pool.
+// Device represents one individual hardware instance that can be selected based
+// on its attributes.
 type Device struct {
-	// Name is a driver-specific identifier for the device.
-	// +required
-	Name string `json:"name"`
+	// Name is unique identifier among all devices managed by
+	// the driver on the node. It must be a DNS label.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
 
-	// Attributes contain additional metadata that can be used in
-	// constraints. If an attribute name overlaps with the pool attribute,
-	// the device attribute takes precedence.
-	//
-	// +optional
-	Attributes []Attribute `json:"attributes,omitempty"`
-
-	// SharedResourcesConsumed contains the pooled resources that are
-	// consumed when this device is allocated.
+	// Attributes defines the attributes of this device.
+	// The name of each attribute must be unique.
 	//
+	// +listType=atomic
 	// +optional
-	SharedResourcesConsumed map[string]resource.Quantity `json:"sharedResourcesConsumed,omitempty"`
+	Attributes []DeviceAttribute `json:"attributes,omitempty" protobuf:"bytes,3,opt,name=attributes"`
 
-	// ClaimResourcesProvided allows the definition of per-device resources
-	// that can be allocated in a manner similar to standard Kubernetes
-	// resources.
-	//
-	// +optional
-	ClaimResourcesProvided []ResourceCapacity `json:"claimResourcesProvided,omitempty"`
+	// TODO for 1.31: define how to support partitionable devices
 }
 
-type ResourceCapacity struct {
-	// Name is the resource name/type.
-	// +required
-	Name string `json:"name"`
-
-	// Capacity is the total capacity of the named resource.
-	// +required
-	Capacity resource.Quantity `json:"capacity"`
-
-	// BlockSize is the increments in which capacity is consumed. For
-	// example, if you can only allocate memory in 4k pages, then the
-	// block size should be "4Ki". Default is 1.
+// DeviceAttribute is a combination of an attribute name and its value.
+type DeviceAttribute struct {
+	// Name is a unique identifier across all possible attributes of devices.
```

Comment:
Can I offer a reframing?  Some of this depends on what we decide for CEL, but I will speculate and we can revisit when that is done.

"""
Name is a unique identifier for this attribute, which will be referenced in ResourceClaim constraints.

Attributes are defined either by the owner of the specific driver (usually the vendor) or by some 3rd party (e.g. the Kubernetes project).  Because attributes are sometimes compared across devices, a given name is expected to mean the same thing and have the same type on all devices.

Attribute names must be either a C-style identifier (e.g. "the_name") or a DNS subdomain followed by a slash ("/") followed by a C-style identifier (e.g. "example.com/the_name").  Attributes whose name does not include the domain prefix are assumed to be part of the driver's domain.  Attributes defined by 3rd parties must include the domain prefix.
"""

Reply:
Looks good to me. The `<dns domain>/<identifier>` syntax is cleaner.